### PR TITLE
Firebase auth

### PR DESCRIPTION
**Pull request by peytonshetler**
## Summary
Firebase authentication was added so that users may sign in or create an account.  
The Sign in functionality was added to the landing screen, while the new account functionality was added to the registration screen.

#### Characteristics

- [x] Feature completed
- [X] Experimental
- [X] UI affected (screen shots included)

### CODE REVIEW

File: Been-There/Been-There/AppDelegate.swift [hunk 1]
```diff
@@ -16,6 +17,11 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
 
     func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?) -> Bool {
         
+        FirebaseApp.configure()
+        
+        let myDatabase = Database.database().reference()
```
Comment (JZDesign) [inline Been-There/Been-There/AppDelegate.swift]: This seems unrelated to setting up Firebase Auth. What does this pertain to?

File: Been-There/Been-There/Controller/LandingScreen.swift [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+//  Been-There
+//
+//  Created by Peyton Shetler on 3/2/19.
+//  Copyright © 2019 Peyton Shetler. All rights reserved.
+//
+
+import UIKit
+import Firebase
+
+class LandingScreen: UIViewController {
```
Comment (JZDesign) [inline Been-There/Been-There/Controller/LandingScreen.swift]: File renames or major refactors going forward should be managed as separate PR's. The goal is to keep the code reviews small and manageable for the reviewer. There was a study of thousands of developers, PRs, and projects and they found that PR's of less than 300 lines of code were reviewed more accurately and the following releases had less defects and bugs than those resulting of larger PR's.

Don't worry about trying to manage that for this situation. Going forward we'll make it a habit to have the PRs handle small themed things like:

- File renames and other naming changes or small things in one pr
- ViewController visual layouts 1 pr
- Functionality for a feature like Firebase Auth being wired into the ViewController 1 PR

That would make it easier for the reviewer to find what is appropriate for each separate thing.

File: Been-There/Been-There/AppDelegate.swift (hunk 1 above)
Comment (peytonshetler) [reply Been-There/Been-There/AppDelegate.swift]: FirebaseApp.configure() just initializes Firebase in the app, while "myDatabase" is something I forgot to delete from earlier I believe.

File: Been-There/Been-There/AppDelegate.swift [hunk 3]
```diff
@@ -19,9 +19,6 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
         
         FirebaseApp.configure()
         
-        let myDatabase = Database.database().reference()
```
Comment (peytonshetler) [inline Been-There/Been-There/AppDelegate.swift]: Deleted this line.  This was something i was using just to test if the app was connected to Firebase.

File: Been-There/Been-There/AppDelegate.swift [hunk 4]
```diff
@@ -19,9 +19,6 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
...
-        myDatabase.setValue("Database test!")
-
```
Comment (peytonshetler) [inline Been-There/Been-There/AppDelegate.swift]: Deleted all this (Lines 22-24)

File: Been-There/Been-There/Controller/LandingScreen.swift [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+        setupMainTitle()
+        setupRegisterButton()
+        setupSignInButton()
+        setupPasswordTextField()
+        setupEmailTextField()
+        setupEmailUnderline()
+        setupPasswordUnderline()
+    }
+    // MARK: Setup/Constraint Methods
+    func setupRegisterButton() {
```
Comment (JZDesign) [inline Been-There/Been-There/Controller/LandingScreen.swift]: To help organize this code and make it easier to review lets utilize extensions to group like functions together.

In your main file `LandingScreen.swift` you'll keep all of the outlets, actions, and lifecycle functions as they are. What I'd like you to do is to create extension files like `UI+LandingScreen.swift` that holds all of the view related functions. Much like you did with the colors. By using the `+LandingScreen` naming convention, that conveys to the team that this file is a direct extension of `LandingScreen` what precedes the `+` denotes what the file is responsible for.

File: Been-There/Been-There/Controller/RegisterScreen.swift [hunk 6]
```diff
@@ -0,0 +1,305 @@
...
+        textField.placeholder = placeholder
+        textField.isSecureTextEntry = hasSecureEntry
+        textField.translatesAutoresizingMaskIntoConstraints = false
+        textField.textColor = .black
+        textField.font = textField.font?.withSize(22)
+        inputContainer.addSubview(textField)
+        completion(textField)
+    }
+    
+    func setupUnderline(_ underline: UIView, color: UIColor) {
```
Comment (JZDesign) [inline Been-There/Been-There/Controller/RegisterScreen.swift]: If any of these are duplicates of what you have in the other file you could extend `UIView` directly and the function would be available to any view.

File: Been-There/Been-There/Controller/RegisterScreen.swift [hunk 7]
```diff
@@ -0,0 +1,305 @@
...
+    
+    func setupFirstNameTextField() {
+        configureTextFieldStyle(firstNameTextField, "First", false) { (firstNameTextField) in
+            if firstNameTextField != nil {
+                self.setupFirstNameTextFieldConstraints()
+            }
+        }
+    }
+    
+    func setupFirstNameTextFieldConstraints() {
```
Comment (JZDesign) [inline Been-There/Been-There/Controller/RegisterScreen.swift]: It looks like some of these functions could be consolidated.

```
func setUpTextFieldConstraints(_ textField: UITextField, topAnchor: NSLayoutAnchor) {
        textField.widthAnchor.constraint(equalTo: inputContainer.widthAnchor, multiplier: 2/3).isActive = true
        textField.heightAnchor.constraint(equalToConstant: 30).isActive = true
        textField.topAnchor.constraint(equalTo: topAnchor).isActive = true
        textField.centerXAnchor.constraint(equalTo: inputContainer.centerXAnchor).isActive = true
}
```

Then you could call that function like `setUpTextFieldConstraints(firstNameTextField, topAnchor: inputContainer.topAnchor)` and reuse it for any other field that has similar needs.

Comment (JZDesign) [reply Been-There/Been-There/Controller/RegisterScreen.swift]: Try and do this for all similar functions, bonus points if you create theme functions that are accessible from any view.

File: Been-There/Been-There/AppDelegate.swift (hunk 4 above)
Comment (JZDesign) [reply Been-There/Been-There/AppDelegate.swift]: Yes these are the comment types I was talking about. The more the merrier.

File: Been-There/Podfile [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+target 'Been-There' do
+  # Comment the next line if you're not using Swift and don't want to use dynamic frameworks
+  use_frameworks!
+
+  # Pods for Been-There
+pod 'Firebase'
+pod 'FirebaseUI/Auth'
+pod 'FirebaseUI/Email'
+pod 'Firebase/Auth'
+pod 'Firebase/Database'
```
Comment (JZDesign) [inline Been-There/Podfile]: If I remember correctly, You should only need a few of these for this ticket.

File: Been-There/Been-There/Controller/RegisterScreen.swift [hunk 9]
```diff
@@ -0,0 +1,305 @@
...
+            }
+            let values = ["firstName": firstName, "lastName": lastName, "email": email]
+            self.registerUserIntoDatabaseWithUID(uid, values: values as [String: AnyObject])
+            print("User created!")  // Successful creation of new user.  This print is a placeholder
+        })
+    }
+    
+    func registerUserIntoDatabaseWithUID(_ uid: String, values: [String: AnyObject]) {
+        let ref = Database.database().reference()
+        let usersReference = ref.child("users").child(uid)
```
Comment (JZDesign) [inline Been-There/Been-There/Controller/RegisterScreen.swift]: It looks like you got ahead of the ticket and started working on the Firebase RTDB. In Kanban we aim to keep the work small and focus only on just the ticket at hand. Don't worry about making changes here. In the future think "does that apply to this ticket?" and if not don't do it.

I'm sure you were just making sure that Auth worked properly, there was a different way to verify that. If you'd like to see that I'll show you later.

File: Been-There/Been-There/Controller/LandingScreen.swift (hunk 5 above)
Comment (JZDesign) [reply Been-There/Been-There/Controller/LandingScreen.swift]: _Spoiler Alert_ This advice is good, but some of or all of it may change based on comments below.

File: Been-There/Podfile (hunk 8 above)
Comment (JZDesign) [reply Been-There/Podfile]: Don't delete any Pod stuff, it's a nightmare. But in the future, try to only work on what is needed for the ticket.

File: Been-There/Been-There/Controller/RegisterScreen.swift (hunk 7 above)
Comment (JZDesign) [reply Been-There/Been-There/Controller/RegisterScreen.swift]: Also, to save even more code, you may consider a StackView.

File: Been-There/Been-There/Controller/RegisterScreen.swift [hunk 10]
```diff
@@ -236,18 +195,6 @@ class RegisterScreen: UIViewController {
...
-        setupPasswordUnderlineConstraints()
-    }
-    
-    func setupPasswordUnderlineConstraints() {
-        passwordUnderline.widthAnchor.constraint(equalTo: passwordTextField.widthAnchor).isActive = true
-        passwordUnderline.heightAnchor.constraint(equalToConstant: 1).isActive = true
-        passwordUnderline.topAnchor.constraint(equalTo: passwordTextField.bottomAnchor).isActive = true
-        passwordUnderline.centerXAnchor.constraint(equalTo: inputContainer.centerXAnchor).isActive = true
-    }
-
```
Comment (peytonshetler) [inline Been-There/Been-There/Controller/RegisterScreen.swift]: The textfields and underlines are both white, making the underlines sloppy and really unnecessary. Therefore, delete.

Comment (JZDesign) [reply Been-There/Been-There/Controller/RegisterScreen.swift]: Good call. And perfect use of a commit.

File: Been-There/Been-There/Controller/RegisterScreen.swift [hunk 11]
```diff
@@ -106,6 +103,31 @@ class RegisterScreen: UIViewController {
...
+        setupInputStackViewConstraints()
+    }
+
+    func setupInputStackViewConstraints() {
+        inputStackView.widthAnchor.constraint(equalTo: inputContainer.widthAnchor).isActive = true
+        inputStackView.heightAnchor.constraint(equalTo: inputContainer.heightAnchor, multiplier: 2/3).isActive = true
+        inputStackView.centerXAnchor.constraint(equalTo: inputContainer.centerXAnchor).isActive = true
+        inputStackView.topAnchor.constraint(equalTo: inputContainer.topAnchor).isActive = true
+    }
+
```
Comment (peytonshetler) [inline Been-There/Been-There/Controller/RegisterScreen.swift]: I converted my inputs from the register screen (first name, last name, email, etc) into a stack view.  The old code is just commented out for now.

Comment (JZDesign) [reply Been-There/Been-There/Controller/RegisterScreen.swift]: Go ahead and create another commit that deletes all of the comments.

File: Been-There/Been-There/Controller/UI+RegisterScreen.swift [hunk 12]
```diff
@@ -0,0 +1,184 @@
...
+        textField.layer.cornerRadius = 5
+        textField.placeholder = placeholder
+        textField.isSecureTextEntry = hasSecureEntry
+        textField.translatesAutoresizingMaskIntoConstraints = false
+        textField.textColor = .black
+        textField.font = textField.font?.withSize(22)
+        completion(textField)
+    }
+    
+    func configureTextFieldConstraints(_ textField: UITextField?) {
```
Comment (JZDesign) [inline Been-There/Been-There/Controller/UI+RegisterScreen.swift]: Because you're doing a null check before you call this function, the `UITextField` parameter doesn't need to be optional. 

Or you could keep it there and do some functional programming. For example:

```
    func setupFirstNameTextField() {
        configureTextFieldStyle(firstNameTextField, "First", false, configureTextFieldConstraints(firstNameTextField)
    }
```

That syntax may be off but you should be able to do something like that. If it doesn't work off the bat lookup `typealias`s and passing closures as parameters.

**Comment by peytonshetler** [comment]
Hallelujah

File: Been-There/Been-There/Controller/UI+RegisterScreen.swift [hunk 13]
```diff
@@ -113,54 +113,32 @@ extension RegisterScreen {
...
     }
     
     func setupFirstNameTextField() {
-        configureTextFieldStyle(firstNameTextField, "First", false) { (firstNameTextField) in
-            if firstNameTextField != nil {
-                self.configureTextFieldConstraints(firstNameTextField)
-            }
-        }
+        configureTextFieldStyle(firstNameTextField, "First", false)
     }
```
Comment (peytonshetler) [inline Been-There/Been-There/Controller/UI+RegisterScreen.swift]: Since I added the width/height constraints to the larger configureTextFieldStyle function, we don't need to add anymore code after the fact therefore we don't need a completion handler anymore so I went ahead and deleted it from the function declaration.
